scala/dllib/src/main/scala/com/intel/analytics/bigdl/dllib/keras/models/Topology.scala

```diff
@@ -472,6 +473,70 @@ abstract class KerasNet[T](implicit val tag: ClassTag[T], implicit val ev: Tenso
     this.fit(x, batchSize, nbEpoch, null)
   }
 
+  def unwrapVectorAsNecessary(colType: DataType): (Row, Int) => Any = {
+    // to support both ML Vector and MLlib Vector
+    if (colType.typeName.contains("vector")) {
+      (row: Row, index: Int) => getVectorSeq(row, colType, index)
+    } else {
+      (row: Row, index: Int) => row.get(index)
+    }
+  }
+
+  private def getDataSet(
+      dataFrame: DataFrame,
+      batchSize: Int,
+      featureCol: String,
+      labelCol: String): FeatureSet[MiniBatch[T]] = {
+
+    val sp = FeatureLabelPreprocessing(SeqToTensor(), ScalarToTensor())
+      .asInstanceOf[Preprocessing[(Any, Option[Any]), Sample[T]]]
+    val featureColIndex = dataFrame.schema.fieldIndex(featureCol)
+    val featureType = dataFrame.schema(featureCol).dataType
+    val featureFunc = unwrapVectorAsNecessary(featureType)
+
+    val labelFunc: (Row) => Option[Any] = if (dataFrame.columns.contains(labelCol)) {
+      val lci = dataFrame.schema.fieldIndex(labelCol)
+      val labelFunc = unwrapVectorAsNecessary(dataFrame.schema(labelCol).dataType)
+      (row: Row) => Some(labelFunc(row, lci))
+    } else {
+      (row: Row) => None
+    }
+
+    val featureAndLabel = dataFrame.rdd.map { row =>
+      val features = featureFunc(row, featureColIndex)
+      val labels = labelFunc(row)
+      (features, labels)
+    }
+
+    val initialDataSet = FeatureSet.rdd(featureAndLabel).transform(sp)
+
+    initialDataSet.transform(SampleToMiniBatch[T](batchSize))
+  }
+
+  def fit(
+      x: DataFrame,
+      batchSize: Int,
+      nbEpoch: Int,
+      featureCol: String,
+      labelCol: String,
+      valX: DataFrame)(implicit ev: TensorNumeric[T]): Unit = {
+    val trainingData = getDataSet(x, batchSize, featureCol, labelCol).toDataSet()
+    val valData = if (valX != null) {
+      getDataSet(valX, batchSize, featureCol, labelCol).toDataSet()
+    } else null
+
+    this.fit(trainingData, nbEpoch, valData)
+  }
+
+  def fit(
+      x: DataFrame,
+      batchSize: Int,
+      nbEpoch: Int,
+      featureCol: String,
+      labelCol: String)(implicit ev: TensorNumeric[T]): Unit = {
```

Comment:
both feature and label columns should be a list of string, similar to https://github.com/intel-analytics/BigDL/blob/branch-2.0/python/orca/src/bigdl/orca/learn/pytorch/pytorch_spark_estimator.py#L104